rustc_codegen_ssa: Make upstream monomorphizations representation sparse

osiewicz opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/149313)*
<!-- homu-ignore:end -->

Upstream monomorphisations are a blessing and a curse of Zed's build performance.

On one hand, we do benefit from it, as builds with share-generics disabled are slower for us (plus we don't really want to use nightly anyways).
On the other, deserializing query results takes *a lot* of time. For some crates close to the end of our compilation pipeline, it's over 400ms per crate.

To make matters worse, I've measured a hit ratio of upstream generics. A sample of such measurement goes as follows:
```
upstream_monomorphization returned None for 28501 distinct monomorphizations.
upstream_monomorphization returned Some 2518 times. Results came from 163 distinct CrateNums.
In total, there were 619731 instantiations to chose from (in upstream_monomorphizations() map))
```
This is horrid for us, as we're using a very small percentage of the map that we spend so much time deserializing from.

This commit tries to (rather clumsily) move us towards a sparse representation of upstream_monomorphizations. Instead of storing <DefId, (GenericArgsRef<'_>)> which is rather heavy to deserialize, we'll resort to storing Hashes of Instances. I plan to make this more foolproof, hence this commit is marked as WIP.
For one, we should probably keep the projection queries. Also, it might be worthwhile to store index pointing at entry within  exported_generics of target crate in order to remedy a potential for collisions.

This commit reduces a `touch crates/editor/src/editor.rs` scenario in Zed for me from 14.5s to 11s.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

osiewicz commented:
That's worth a try I guess. 
It seems to only support encoding data with fixed size, which would not work with GenericArgsRef I think?

rust-timer commented:
Queued 4fcb6e64fedb642e69a609c131cb6bd21443a776 with parent 7934bbdf84a6b9a30297caf4f4f38286dedf876a, future [comparison URL](https://perf.rust-lang.org/compare.html?start=7934bbdf84a6b9a30297caf4f4f38286dedf876a&end=4fcb6e64fedb642e69a609c131cb6bd21443a776).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.6 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

osiewicz commented:
Fair disclosure: I have used Claude Opus to generate code submitted in this PR.

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/21847602844/job/63047084658) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/63047084658) [(plain)](https://github.com/rust-lang/rust/commit/1b2e745cbdd6efa31edb08fc6c11f128f9bd43d1/checks/63047084658/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
fmt check
Diff in /checkout/compiler/rustc_metadata/src/rmeta/encoder.rs:702:
             });
 
         let exported_generic_symbol_hashes = stat!("exported-symbol-hashes", || {
-            self.encode_exported_generic_symbol_hashes(
-                tcx.exported_generic_symbols(LOCAL_CRATE),
-            )
+            self.encode_exported_generic_symbol_hashes(tcx.exported_generic_symbols(LOCAL_CRATE))
         });
 
         // Encode the hygiene data.
Diff in /checkout/compiler/rustc_metadata/src/rmeta/encoder.rs:2269:
         &mut self,
         exported_symbols: &[(ExportedSymbol<'tcx>, SymbolExportInfo)],
     ) -> LazyValue<exported_symbol_hash_map::ExportedSymbolHashTableRef> {
-        use rustc_data_structures::stable_hasher::{HashStable, StableHasher};
         use exported_symbol_hash_map::{ExportedSymbolHashTable, ExportedSymbolHashTableRef};
+        use rustc_data_structures::stable_hasher::{HashStable, StableHasher};
 
         let tcx = self.tcx;
         let mut table = ExportedSymbolHashTable::with_capacity(exported_symbols.len(), 87);
Diff in /checkout/compiler/rustc_metadata/src/rmeta/mod.rs:4:
 use decoder::LazyDecoder;
 pub(crate) use decoder::{CrateMetadata, CrateNumMap, MetadataBlob, TargetModifiers};
 use def_path_hash_map::DefPathHashMapRef;
-use exported_symbol_hash_map::ExportedSymbolHashTableRef;
 use encoder::EncodeContext;
 pub use encoder::{EncodedMetadata, encode_metadata, rendered_const};
+use exported_symbol_hash_map::ExportedSymbolHashTableRef;
 pub(crate) use parameterized::ParameterizedOverTcx;
 use rustc_abi::{FieldIdx, ReprOptions, VariantIdx};
 use rustc_data_structures::fx::FxHashMap;
Diff in /checkout/compiler/rustc_codegen_ssa/src/back/symbol_export.rs:507:
     providers.queries.is_unreachable_local_definition = is_unreachable_local_definition_provider;
     providers.queries.upstream_drop_glue_for = upstream_drop_glue_for_provider;
     providers.queries.upstream_async_drop_glue_for = upstream_async_drop_glue_for_provider;
-    providers.queries.upstream_monomorphization_hashes =
-        upstream_monomorphization_hashes_provider;
+    providers.queries.upstream_monomorphization_hashes = upstream_monomorphization_hashes_provider;
     providers.queries.upstream_monomorphization_for_hash =
         upstream_monomorphization_for_hash_provider;
     providers.queries.wasm_import_module_map = wasm_import_module_map;
Diff in /checkout/compiler/rustc_middle/src/middle/exported_symbols.rs:63:
     /// required lang items are unavailable.
     pub fn to_instance(&self, tcx: TyCtxt<'tcx>) -> Option<Instance<'tcx>> {
         match *self {
-            ExportedSymbol::Generic(def_id, args) => Some(Instance {
-                def: InstanceKind::Item(def_id),
-                args,
-            }),
+            ExportedSymbol::Generic(def_id, args) => {
+                Some(Instance { def: InstanceKind::Item(def_id), args })
+            }
             ExportedSymbol::DropGlue(ty) => {
                 let def_id = tcx.lang_items().drop_in_place_fn()?;
                 Some(Instance {
fmt: checked 6697 files
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:00:50
  local time: Tue Feb 10 01:18:35 UTC 2026
```

</details>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/21847761677/job/63047533464) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/63047533464) [(plain)](https://github.com/rust-lang/rust/commit/9c3a78167935756477b0c807c13ed23fe5902953/checks/63047533464/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
fmt: checked 6697 files
tidy check
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
tidy [alphabetical (compiler)]: /checkout/compiler/rustc_metadata/src/rmeta/parameterized.rs:71: line not in alphabetical order (tip: use --bless to sort this list)
tidy [alphabetical (compiler)]: FAIL
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.3)
Collecting pip
---
info: ES-Check: there were no ES version matching errors!  🎉
typechecking javascript files
tidy: The following check failed: alphabetical (compiler)
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy --root-path=/checkout --cargo-path=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo --output-dir=/checkout/obj/build --concurrency=4 --npm-path=/node/bin/yarn --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1612:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1365:29

Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:02:55
  local time: Tue Feb 10 01:27:09 UTC 2026
  network time: Tue, 10 Feb 2026 01:27:09 GMT
##[error]Process completed with exit code 1.
##[group]Run echo "disk usage:"
```

</details>

cjgillot commented on compiler/rustc_middle/src/ty/instance.rs:
```diff
@@ -225,15 +227,27 @@ impl<'tcx> Instance<'tcx> {
-            InstanceKind::Item(def) => tcx
-                .upstream_monomorphizations_for(def)
-                .and_then(|monos| monos.get(&self.args).cloned()),
-            InstanceKind::DropGlue(_, Some(_)) => tcx.upstream_drop_glue_for(self.args),
```
Is this query still used anywhere?

cjgillot reviewed (commented):
IIUC, most of the gains from this PR avoid decoding `exported_generic_symbols`. Is that so?

Do you have more detailed statistics about cases where `upstream_monomorphization` returns `None`? Typical structure of the instances?

This PR introduces a heavy machinery to filter calls to `exported_generic_symbols`. Did you try manual filters?

For instance, early return `None` if the generic arguments contain a type from the current crate?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149313/f21b4c0888ca6a388fcf9365fc0044fa0642ab98..67cb41fc5729ccdbddd6a874017f52a6c5d3dc9f)*

osiewicz committed:
rustc_codegen_ssa: Make upstream monomorphizations representation sparse
Upstream monomorphisations are a blessing and a curse of Zed's build performance.

On one hand, we do benefit from it, as builds with share-generics disabled are slower for us (plus we don't really want to use nightly anyways).
On the other, deserializing query results takes *a lot* of time. For some crates close to the end of our compilation pipeline, it's over 400ms per crate.

To make matters worse, I've measured a hit ratio of upstream generics. A sample of such measurement goes as follows:
```
upstream_monomorphization returned None for 28501 distinct monomorphizations.
upstream_monomorphization returned Some 2518 times. Results came from 163 distinct CrateNums.
In total, there are 619731 instantiations
```
This is horrid for us, as we're using a very small percentage of the map that we spend so much time deserializing from.

This commit tries to (rather clumsily) move us towards a sparse representation of upstream_monomorphizations. Instead of storing <DefId, (GenericArgsRef<'_>)> which is rather heavy to deserialize, we'll resort to storing Hashes of Instances.

This commit reduces a `touch crates/editor/src/editor.rs` scenario in Zed for me from 14.5s to 11s.

osiewicz committed:
Remove legacy upstream_monomorphization queries

upstream_monomorphizatons required loading all exported symbols of all dependencies whenever any of them changed. This is problematic for large codebases like Zed.

osiewicz commented:
I've removed legacy `upstream_monomorphizations` queries. This PR should reduce the # of exported_symbols we need to look at from `O(all_deps)` to `O(crates_containing_usable_monos)`, which in case of Zed should reduce # of crates we need to look at about 14-fold.

I am still a bit lost with projection queries, so any and all advice is most welcome. We'd love to have it in 1.97.

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25703134578/job/75467650573) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/75467650573) [(plain)](https://github.com/rust-lang/rust/commit/dd7a79f96c95361c1239fceb1e14ed71f0dbd00e/checks/75467650573/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 use rustc_hir as hir;
Diff in /checkout/compiler/rustc_middle/src/ty/instance.rs:225:
 
         match self.def {
             InstanceKind::Item(_)
-             | InstanceKind::DropGlue(_, Some(_))
-             | InstanceKind::AsyncDropGlueCtorShim(_, _) => {
-                 let hash: Hash128 = tcx.with_stable_hashing_context(|mut hcx| {
-                     let mut hasher = StableHasher::new();
-                     self.stable_hash(&mut hcx, &mut hasher);
-                     hasher.finish()
-                 });
-                 let candidates = tcx.upstream_monomorphization_for_hash(hash)?;
-                 // Candidates are sorted by StableCrateId. First verified match wins.
-                 for &cnum in candidates.iter() {
-                     for (sym, _) in tcx.exported_generic_symbols(cnum).iter() {
-                         if sym.to_instance(tcx).as_ref() == Some(self) {
-                             return Some(cnum);
-                         }
-                     }
-                 }
-                 None
-             }
+            | InstanceKind::DropGlue(_, Some(_))
+            | InstanceKind::AsyncDropGlueCtorShim(_, _) => {
+                let hash: Hash128 = tcx.with_stable_hashing_context(|mut hcx| {
+                    let mut hasher = StableHasher::new();
+                    self.stable_hash(&mut hcx, &mut hasher);
+                    hasher.finish()
+                });
+                let candidates = tcx.upstream_monomorphization_for_hash(hash)?;
+                // Candidates are sorted by StableCrateId. First verified match wins.
+                for &cnum in candidates.iter() {
+                    for (sym, _) in tcx.exported_generic_symbols(cnum).iter() {
+                        if sym.to_instance(tcx).as_ref() == Some(self) {
+                            return Some(cnum);
+                        }
+                    }
+                }
+                None
+            }
             InstanceKind::AsyncDropGlue(_, _) => None,
             InstanceKind::FutureDropPollShim(_, _, _) => None,
 
Diff in /checkout/compiler/rustc_middle/src/queries.rs:61:
 use rustc_data_structures::svh::Svh;
 use rustc_data_structures::unord::{UnordMap, UnordSet};
 use rustc_errors::{ErrorGuaranteed, catch_fatal_errors};
+use rustc_hashes::Hash128;
 use rustc_hir as hir;
 use rustc_hir::attrs::{EiiDecl, EiiImpl, StrippedCfgItem};
 use rustc_hir::def::{DefKind, DocLinkResMap};
Diff in /checkout/compiler/rustc_middle/src/queries.rs:69:
 use rustc_hir::{ItemLocalId, ItemLocalMap, PreciseCapturingArgKind, TraitCandidate};
 use rustc_index::IndexVec;
 use rustc_lint_defs::LintId;
-use rustc_hashes::Hash128;
 use rustc_macros::rustc_queries;
 use rustc_session::Limits;
 use rustc_session::config::{EntryFnType, OptLevel, OutputFilenames, SymbolManglingVersion};
Diff in /checkout/compiler/rustc_codegen_ssa/src/back/symbol_export.rs:404:
 ) -> FxIndexMap<Hash128, smallvec::SmallVec<[CrateNum; 1]>> {
     let mut map: FxIndexMap<Hash128, smallvec::SmallVec<[CrateNum; 1]>> = Default::default();
     for &cnum in tcx.crates(()) {
-    for &hash in tcx.exported_generic_symbol_hashes(cnum) {
+        for &hash in tcx.exported_generic_symbol_hashes(cnum) {
             map.entry(hash).or_default().push(cnum);
         }
     }
Diff in /checkout/compiler/rustc_metadata/src/rmeta/mod.rs:279:
     stable_order_of_exportable_impls: LazyArray<(DefIndex, usize)>,
     exported_non_generic_symbols: LazyArray<(ExportedSymbol<'static>, SymbolExportInfo)>,
     exported_generic_symbols: LazyArray<(ExportedSymbol<'static>, SymbolExportInfo)>,
-   exported_generic_symbol_hashes: LazyValue<ExportedSymbolHashTableRef>,
+    exported_generic_symbol_hashes: LazyValue<ExportedSymbolHashTableRef>,
 
     syntax_contexts: SyntaxContextTable,
     expn_data: ExpnDataTable,
Diff in /checkout/compiler/rustc_metadata/src/rmeta/encoder.rs:772:
                 stable_order_of_exportable_impls,
---
                 tables,
                 syntax_contexts,
Diff in /checkout/compiler/rustc_metadata/src/rmeta/encoder.rs:2273:
 
         self.lazy_array(exported_symbols.iter().cloned())
     }
-
 
     fn encode_exported_generic_symbol_hashes(
         &mut self,
```

</details>

bjorn3 commented:
Do I understand correctly that enough has changed since the last perf run that another one would be useful?

osiewicz commented:
Yes, perf run would be most welcome, but I'm seeing a regression on Zed itself. I'll dig into that first.
